### PR TITLE
show the latest snapshot of a camera connected to Home Assistant

### PR DESCRIPTION
**Pull request by pfefferle**
#### Description

This PR allows to ask for the latest snapshot of a camera connected to Home Assistant

#### Type of PR
- [ ] Bugfix
- [X] Feature implementation
- [ ] Refactor of code (without functional changes)
- [ ] Documentation improvements
- [ ] Test improvements

#### Testing

Say "Show me the latest snapshot of {Camera Name}."

### CODE REVIEW

**Comment by pfefferle** [comment]
@krisgesling I am not sure if I should also add a sound output.

**Comment by krisgesling** [comment]
maybe `if not self.gui.connected` then we want to say something? 

otherwise the most I'd go for is a `self.acknowledge()` tone

**Comment by pfefferle** [comment]
@krisgesling what about playing that sound if connected or not. I liked the idea and added it also to the "open Home Assistant" command.

**Comment by krisgesling** [comment]
Yeah, I think it's worth us considering what we want the behaviour to be with or without a display connected. There's a few options as I see it:
- act as if a display is connected - and nothing will show because there's no display
- report that no display exists so they won't get the full experience
- provide different experiences if a display exists
- disable display focused intents if there is no gui connected

They last one I think might have merit here. If there's no display, should it even respond to utterances like "show me the latest snapshot"?

On the other hand, does having different functionality on different devices make it more confusing for users? 
eg If I say the same thing to a different Mycroft device with the same Skill installed and it says: "I'm sorry I don't know what that means" - that's kind of weird.

**Comment by pfefferle** [comment]
Now I see! I totally agree! To be honest, I never thought about the case that there might be no display connected (and don't ask what I thought `self.gui.connected` is checking)!

I would go with

> report that no display exists so they won't get the full experience

and add an error message for that.

I will also add it to "open HomeAssistant"!

**Comment by Tony763** [comment]
I do not have display so from my point of view I would rather wanted to have intents which use booth to respond only with voice without info about no display. And intents that are only for gui to respond with info message that display is not connected.

**Comment by pfefferle** [comment]
@krisgesling @Tony763 I added an error if there is no gui available

**Comment by Tony763** [comment]
I will send translations tomorrow.

**Comment by pfefferle** [comment]
@krisgesling can you please re-review?

File: __init__.py
Comment (krisgesling) [inline __init__.py]: Does this get used? 

I'm guessing the intention is to allow the `_handle_camera_actions()` method to be extended in the future to handle different types of actions?

If there's only one Action type at the moment, maybe we rename the method to specify what it does now eg `_show_camera_image()`. Then in the future we can call it from a more generic camera actions handler?

File: __init__.py
Comment (krisgesling) [inline __init__.py]: Is the idea that there will only be a single `camera.intent` or is there likely to be more?
If there will be more, then we should consider renaming this to represent the specifics of this intent a little more, eg `camera.show.image.intent`?

**Review by krisgesling — COMMENTED** [review]
Hey @pfefferle sorry for the delay - overall it looks good. 

A few suggestions on naming below that I think will help set the foundation for future camera functionality to be added more easily.

**Comment by krisgesling** [comment]
Oh, it would be great if we could add a test for this too

File: __init__.py
Comment (pfefferle) [reply __init__.py]: I used `show.camera.image.intent` to be consistent with the other namings of the plugin.

**Comment by pfefferle** [comment]
@Tony763 can you support me by testing the camera GUI response?

**Comment by Tony763** [comment]
@pfefferle I could try to write some tests, but I can't test gui as I run my HA and Mycroft on my headless intel server. I will look for some way to use RDP or similar protocol to stream video output.

**Comment by pfefferle** [comment]
@Tony763 @krisgesling the test will also run in headless mode doesn't it? Is it even possible to test for GUI output? so should I only test for the error case?

**Comment by Tony763** [comment]
In headless but give me a week or two, I will think about it ;)

út 7. 12. 2021 v 6:52 odesílatel Matthias Pfefferle <
***@***.***> napsal:

**Comment by krisgesling** [comment]
Currently there's no built-in tests to guarantee that it's actually rendered. However intent clashes are the bigger concern, so it would be good to check that the intent handler triggers for what it's meant to and not things like:
> show me a picture of a cat

**Comment by pfefferle** [comment]
@krisgesling makes sense.

@Tony763 I added some basic tests, but the GitHub Action seems to be broken on my fork, any ideas? https://github.com/pfefferle/skill-homeassistant/runs/4463932766?check_suite_focus=true

**Comment by Tony763** [comment]
Yes, there was a breaking change in mycroft-core installation, fix for it is in PR #89.

**Comment by pfefferle** [comment]
I added tests, to check for an existing and non existing entity.

**Comment by pfefferle** [comment]
@krisgesling can you re-check?

**Comment by pfefferle** [comment]
@Tony763 @krisgesling this is weird... it has no issues on my fork https://github.com/pfefferle/skill-homeassistant/runs/4481764469

**Comment by Tony763** [comment]
It's okay, workflow change is not merged to HA repo yet, so it's fail there. Workflow on PR is always used a one from repository into which you sended a PR. It's a security feature. So you are unable to run malicious code before approve. When my change is merged, this will also work there.

**Comment by Tony763** [comment]
Pylint is not working as it is not installed. :slightly_smiling_face:  In Your repo its okay.

Wait for @krisgesling to check my PR and merge. :slightly_smiling_face: 

image

**Comment by Tony763** [comment]
Hi @pfefferle, to incorporate (activate) latest change in workflow to this PR, You have to rebase it with latest commit made to branch 20.08. Ping me if You need help with commands :slightly_smiling_face:

File: __init__.py
Comment (Tony763) [inline __init__.py]: Hi @pfefferle, would You add also type annotations as I did in my last PR so we keep same style?

```
def handle_show_camera_image_intent(self, message: Message) -> None:
       """Handle show camera image intent."""
```

Comment (pfefferle) [reply __init__.py]: sure!

**Comment by pfefferle** [comment]
@Tony763 @krisgesling should be ready to review again :)

File: __init__.py
Comment (krisgesling) [inline __init__.py:333]: I'm assuming you've been testing this on an actual system. This just jumped out as a possible error point.

Does the client `url` or the `entity_picture` always have a `/` included?
I see in the test config the mock `entity_picture` has a `/` but it's not what I'd necessarily expect.

**Review by krisgesling — APPROVED** [review]
Thanks Matthias!

Assuming the below comment is not an issue this is ready to be merged :tada:

File: __init__.py
Comment (pfefferle) [reply __init__.py:333]: `entity_picture` is a generated API URL (by Home Assistiant) to the image and is always an absolute path that starts always with a `/` and the URL should have no trailing slash, because it is generated using the IP/Domain and the Port: https://github.com/MycroftAI/skill-homeassistant/blob/20.08/ha_client.py#L88

Example attributes of a camera:

```
access_token: token_123
frontend_stream_type: hls
entity_picture: /api/camera_proxy/camera.camera_name?token=token_123
friendly_name: Camera Friendly Name
```

So I think there is no problem.

Comment (pfefferle) [reply __init__.py:333]: and welcome back @krisgesling :)

Comment (krisgesling) [reply __init__.py:333]: Excellent - thanks for clarifying and for the warm return :smile:

**Pull request merged**
